### ISSUE TITLE
Rule proposal: `prefer-named-import`

### ISSUE THREAD

**Issue by B2o5T** [sindresorhus/eslint-plugin-unicorn#2146]
### Description

I don't know if I chose right repo to suggest this rule or if I should ask it in `eslint-plugin-import` instead.

as @fisker said https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2087#issuecomment-1543791405 development experience in `eslint-plugin-import` is bad and also we can wait for a years for the new good rules there.

Also, I am not sure but I remember that I already saw this rule implemented, but now I can't find it. (I could be wrong too)

### Fail

I think I don't need to mention that named imports/exports are preferable over default.
if module exports named and default export, consider using named import instead default

```js
// foo source code

let foo = '...'

export { foo }
export default foo
```

```js
import foo from 'foo'
```

### Pass

```js
import { foo } from 'foo'
```

**Comment by sindresorhus** [comment]
> I think I don't need to mention that named imports/exports are preferable over default.

I don't think it's common sentiment that named exports are preferred over default ones. Linking to some sources would help back up your argument.

> if module exports named and default export, consider using named import instead default

Exporting the same thing as both named and default is completely wrong. We should definitely have a rule to prevent that.

**Comment by sindresorhus** [comment]
I'm -0 on this. I don't think named exports should always be preferred. There are many cases where a default export makes more sense. However, I won't block this rule from happening, but it would not be in the recommended preset.

**Comment by fisker** [comment]
May be this rule is what you are looking for [`import/no-default-export`](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-default-export.md)

**Comment by B2o5T** [comment]
> I don't think it's common sentiment that named exports are preferred over default ones. Linking to some sources would help back up your argument.

- consistent naming

> If something is an Apple in one file, I shouldn’t call it Orange in another file. An Apple should always be an Apple.

- renaming imports across codebases in ide editors works badly if module is exported as the default export

and much more described in ESLint's author blogpost

https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

> May be this rule is what you are looking for [import/no-default-export](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/no-default-export.md?rgh-link-date=2023-05-26T10%3A15%3A31Z)

it's not exactly the same, my proposal for lining imports for external packages, and not forbidding default exports for internal code files.
If external package export the same thing as both named and default (like `clsx` or `GraphiQL` packages do) then propose to use the named version in case of using default import

**Comment by fisker** [comment]
> If external package export the same thing as both named and default

I believe there is no way to know that unless we actually import the package.

[`import-style`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/import-style.md) should fit your needs?

**Comment by silverwind** [comment]
CJS modules like [webpack](https://github.com/webpack/webpack/issues/13098) can not properly be statically analyze by node's [cjs-module-lexer](https://github.com/nodejs/cjs-module-lexer), which makes named imports from them impossible, so the rule would be a very complex one to be able to detect such scenarios.

**Comment by B2o5T** [comment]
I found what was a searched rule [import/no-named-as-default](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-named-as-default.md) but he warns only on local files

<img width="709" alt="image" src="https://github.com/sindresorhus/eslint-plugin-unicorn/assets/7361780/e24341ce-904b-41d9-82fe-3cb5a438ee2e">
<img width="252" alt="image" src="https://github.com/sindresorhus/eslint-plugin-unicorn/assets/7361780/cbe10ea7-0187-406a-a184-dcbe50d88336">

@silverwind is right to let's close here and suggest checking external packages in `eslint-plugin-import` for `import/no-named-as-default`!

Thanks for everybody 😜
